libs/libcommon/src/libcommon/queue.py

```diff
@@ -609,17 +547,70 @@ def get_next_waiting_job(
                 )
         raise EmptyQueueError("no job available")
 
-    def _start_job(self, job: Job) -> Job:
+    def _start_newest_job_and_cancel_others(self, job: Job) -> Job:
+        """Start a job (the newest one for unicity_id) and cancel the other ones.
+
+        A lock is used to ensure that the job is not started by another worker.
+
+        Args:
+            job: the job to start
+
+        Returns:
+            the started job
+
+        Raises:
+            AlreadyStartedError: if the job is already started by another worker.
+            LockTimeoutError: if the lock could not be acquired after 20 retries.
+        """
         # could be a method of Job
-        job.update(started_at=get_datetime(), status=Status.STARTED)
-        return job
+        RETRIES = 20
+        try:
+            # retry for 2 seconds
+            with lock(key=job.unicity_id, owner=str(job.pk), sleeps=[0.1] * RETRIES):
+                # get all the pending jobs for the same unicity_id
+                waiting_jobs = Job.objects(
+                    unicity_id=job.unicity_id, status__in=[Status.WAITING, Status.STARTED]
+                ).order_by("-created_at")
+                datetime = get_datetime()
+                # raise if any job has already been started for unicity_id
+                num_started_jobs = waiting_jobs(status=Status.STARTED).count()
+                if num_started_jobs > 0:
+                    if num_started_jobs > 1:
+                        logging.critical(f"job {job.unicity_id} has been started {num_started_jobs} times. Max is 1.")
+                    raise AlreadyStartedError(f"job {job.unicity_id} has been started by another worker")
+                # get the most recent one
+                first_job = waiting_jobs.first()
```

Comment:
What if the first job is a job that is based on a parent job that has not been run yet ? It might run on outdated data no ?

Reply:
Yes, it happens a lot, and it's not an issue. It's better to have an explicit error or a partial response than an empty cache. And it will be re-run when the parent job is finished.

Anyway, this PR does not change anything in that aspect.

Reply:
ah yes indeed !